compose: Automatically save drafts while typing

nirvedh-harpal opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #36102
Fixes: [#36762](https://github.com/zulip/zulip/issues/36762)

This PR improves the compose box reliability by adding an auto-save mechanism that ensures drafts are saved automatically while the user is typing. This helps prevent message loss in scenarios such as bad network connections, accidental tab closures, or power loss.

## Implementation

Added throttled auto-save for compose drafts that triggers at most once every 60 seconds while the user is actively composing. The implementation uses Lodash's `throttle` function with `leading: false` and `trailing: true` to ensure the draft is saved after the user pauses typing.

**Key changes:**

1. **New `throttled_update_draft` function** (`web/src/compose.ts`):
   - Wraps `drafts.update_draft()` with a 60-second throttle
   - Configured with `leading: false` (don't save immediately on first input)
   - Configured with `trailing: true` (save once after typing pauses)

2. **Auto-save trigger** (`web/src/compose.ts`):
   - Called `throttled_update_draft()` in the compose textarea input event handler
   - Ensures drafts are created/updated automatically while typing

3. **Fix for draft recreation** (`web/src/drafts.ts`):
   - When a draft is cleared (content completely deleted), reset `compose_draft_id` to `undefined`
   - This forces creation of a new draft if the user starts typing again
   - Fixes existing bug where drafts weren't created after clearing content, switching tabs, and returning

## Testing

Manually tested the following scenarios:

- ✅ Start typing a message → verify draft is created after 60 seconds (for messages > 2 characters)
- ✅ Continue editing → verify draft updates at most once per 60 seconds
- ✅ Clear message completely and start typing again → verify new draft is created
- ✅ Clear message, switch to another browser tab, return and start typing → verify new draft is created (this previously failed)
- ✅ Type, wait for auto-save, then send message → verify draft is cleaned up properly

## Demo

<details>
<summary>Auto-save demonstration</summary>

https://github.com/user-attachments/assets/6ea78055-d37e-4579-8ba4-3398dc62001d

**Note:** The video shows a 5-second demo for brevity. In production, the auto-save interval is 60 seconds.

</details>

## Discussion

CZO thread: [#issues > 📂 save drafts more?](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20save.20drafts.20more.3F/near/2262154)

Related PR that introduced the original draft clear bug: #35444

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?

alya commented:
@evykassirer Are you up for reviewing this one?

evykassirer commented:
There are now two PRs open for this issue, this one and #36389

evykassirer commented on web/src/drafts.ts:
```diff
@@ -444,10 +444,12 @@ export let update_draft = (opts: UpdateDraftOptions = {}): string | undefined =>
+        // a fresh draft.
         if (draft_id) {
             draft_model.deleteDrafts([draft_id]);
+            compose_draft_id = undefined;
```
This is a great fix, but seems unrelated to the other changes. Can you open a separate PR for it?

timabbott commented on web/src/compose_setup.ts:
```diff
@@ -57,6 +57,14 @@ function setup_compose_actions_hooks(): void {
+    },
+    60000,
+    {leading: false, trailing: true},
+);
```
I concern with this API is that you're not actually checking that the compose box hasn't closed and moved on to a different pursuit in the meantime. I think a clean method would look at the current `draft_id` for the compose box, compare that with what was passed into this function, and return early if that draft is already gone.

Doing that correctly may be a bit tricky, though, since the throttling function may not know how to specialize itself with the draft ID.

timabbott replied:
Posted in [#issues > 📂 save drafts more? @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20save.20drafts.20more.3F/near/2305316). My current thinking is that probably this is the right code and we should just have a comment making the argument I posted in CZO.

nirvedh-harpal committed:
Automatically save drafts while typing.

Added throttled auto-save for compose drafts that triggers at most once every 60 seconds while the user is actively composing. The implementation uses Lodash's throttle function with leading: false and trailing: true to ensure the draft is saved after the user pauses typing.

Fixes: #36102

nirvedh-harpal committed:
Reset draft ID when the content is cleared

Fixes existing bug where drafts weren't created after clearing content, switching tabs, and returning.

When a draft is cleared (content completely deleted), reset compose_draft_id to undefined
This forces creation of a new draft if the user starts typing again

Fixes: #36762

nirvedh-harpal commented on web/src/compose_setup.ts:
```diff
@@ -109,6 +117,9 @@ export function initialize(): void {
             compose_state.set_is_content_unedited_restored_draft(false);
         }
+
+        // Save the draft, but throttled to at most once per minute.
```
Updated the comment.
